consensus/istanbul/backend/engine.go

```diff
@@ -416,19 +418,25 @@ func (sb *Backend) UpdateValSetDiff(chain consensus.ChainReader, header *types.H
 	return writeValidatorSetDiff(header, []istanbul.ValidatorData{}, []istanbul.ValidatorData{})
 }
 
-// Returns whether or not a particular header represents the last block in the epoch.
+// IsLastBlockOfEpoch returns whether or not a particular header represents the last block in the epoch.
 func (sb *Backend) IsLastBlockOfEpoch(header *types.Header) bool {
 	return istanbul.IsLastBlockOfEpoch(header.Number.Uint64(), sb.config.Epoch)
 }
 
-// Returns the size of epochs in blocks.
+// EpochSize returns the size of epochs in blocks.
 func (sb *Backend) EpochSize() uint64 {
 	return sb.config.Epoch
 }
 
-// Returns the size of the lookback window for calculating uptime (in blocks)
-func (sb *Backend) LookbackWindow() uint64 {
-	return sb.config.LookbackWindow
+// LookbackWindow returns the size of the lookback window for calculating uptime (in blocks)
+// Value is constant during an epoch
+func (sb *Backend) LookbackWindow(header *types.Header, state *state.StateDB) uint64 {
+	return uptime.ComputeLookbackWindow(
+		sb.config.Epoch,
+		sb.config.DefaultLookbackWindow,
+		sb.chain.Config().IsDonut(header.Number),
```

Comment:
I kind of think this should use `IsDonut(firstBlockOfEpoch)`.  Otherwise, when Donut activates, we could end up with different values within that epoch in a scenario where the value in the contract has by then already been modified through a governance proposal (or initialized to an incorrect value).  That is extremely unlikely to happen, but it still seems more correct to use the beginning of the epoch for the Donut check so that it only takes effect at the beginning of the epoch after activation (or the same epoch if activation is on the epoch's first bock).

Of course, there is a caveat that the CIP says "Starting at HARDFORK_BLOCK, block creation and validation MUST read the lookback window from chain state by calling the BlockchainParameters contract, rather than reading chain paramaters."  But the CIP is actually out of date already, since it doesn't take into account the new contract design where `uptimeLookbackWindow` is a struct rather than `uint256`.

Reply:
good idea.